ws/adapter_test.go

```diff
@@ -0,0 +1,244 @@
+		},
+		{
+			name:     "unsubscribe from channel with unsubscribe set to fail",
+			thingKey: thingKey,
+			chanID:   chanID,
+			subtopic: subTopic,
+			pubsub:   false,
+			fail:     true,
+			err:      ws.ErrFailedUnsubscribe,
+		},
+		{
+			name:     "subscribe to channel with valid thingKey, chanID, subtopic",
+			thingKey: thingKey,
+			chanID:   chanID,
+			subtopic: subTopic,
+			pubsub:   true,
+			fail:     false,
+			err:      nil,
+		},
+		{
+			name:     "subscribe again to channel with valid thingKey, chanID, subtopic",
+			thingKey: thingKey,
+			chanID:   chanID,
+			subtopic: subTopic,
+			pubsub:   true,
+			fail:     false,
+			err:      nil,
+		},
+		{
+			name:     "subscribe to channel with invalid chanID and invalid thingKey",
+			thingKey: "invalid",
+			chanID:   "0",
+			subtopic: subTopic,
+			pubsub:   true,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+		{
+			name:     "subscribe to channel with empty channel",
+			thingKey: thingKey,
+			chanID:   "",
+			subtopic: subTopic,
+			pubsub:   true,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+		{
+			name:     "subscribe to channel with empty thingKey",
+			thingKey: "",
+			chanID:   chanID,
+			subtopic: subTopic,
+			pubsub:   true,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+		{
+			name:     "subscribe to channel with empty thingKey and empty channel",
+			thingKey: "",
+			chanID:   "",
+			subtopic: subTopic,
+			pubsub:   true,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+		{
+			name:     "unsubscribe from channel with valid thingKey, chanID, subtopic",
+			thingKey: thingKey,
+			chanID:   chanID,
+			subtopic: subTopic,
+			pubsub:   false,
+			fail:     false,
+			err:      nil,
+		},
+		{
+			name:     "unsubscribe from channel with valid thingKey, chanID, and empty subtopic",
+			thingKey: thingKey,
+			chanID:   chanID,
+			subtopic: "",
+			pubsub:   false,
+			fail:     false,
+			err:      nil,
+		},
+		{
+			name:     "unsubscribe from channel with empty channel",
+			thingKey: thingKey,
+			chanID:   "",
+			subtopic: subTopic,
+			pubsub:   false,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+		{
+			name:     "unsubscribe from channel with empty thingKey",
+			thingKey: "",
+			chanID:   chanID,
+			subtopic: subTopic,
+			pubsub:   false,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+		{
+			name:     "unsubscribe from channel with empty thingKey and empty channel",
+			thingKey: "",
+			chanID:   "",
+			subtopic: subTopic,
+			pubsub:   false,
+			fail:     false,
+			err:      ws.ErrUnauthorizedAccess,
+		},
+	}
+
+	for _, tt := range cases {
+		pubsub.SetFail(tt.fail)
+		if tt.pubsub == true {
+			t.Run(tt.name, func(t *testing.T) {
+				assert.Equal(t, tt.err, svc.Subscribe(context.Background(), tt.thingKey, tt.chanID, tt.subtopic, c))
+			})
+		} else {
+			t.Run(tt.name, func(t *testing.T) {
```

Comment:
t.Run makes it a little easier to debug. Since all the testcases become individually available, in Testing section of vsCode.
But I'll make this change!